controller/src/main/java/io/pravega/controller/server/bucket/BucketManager.java

```diff
@@ -62,28 +61,36 @@ protected void doStart() {
     }
 
     private CompletableFuture<Void> tryTakeOwnership(int bucket) {
-        return RetryHelper.withIndefiniteRetriesAsync(() -> streamMetadataStore.takeBucketOwnership(bucket, processId, executor),
-                e -> log.warn("exception while attempting to take ownership"), executor)
+        return RetryHelper.withIndefiniteRetriesAsync(() -> bucketStore.takeBucketOwnership(serviceType, bucket, 
+                processId, executor),
+                e -> log.warn("exception while attempting to take ownership for bucket {} for service {}", bucket, serviceType), executor)
                 .thenCompose(isOwner -> {
-                    if (isOwner && buckets.stream().noneMatch(x -> x.getBucketId() == bucket)) {
-                        log.info("Taken ownership for bucket {}", bucket);
-                        StreamCutBucketService bucketService = new StreamCutBucketService(bucket, streamMetadataStore,
-                                streamMetadataTasks, executor, requestTracker);
-                        buckets.add(bucketService);
+                    if (isOwner) {
+                        log.info("Taken ownership for bucket {} for service {}", bucket, serviceType);
+
+                        // Once we have taken ownership of the bucket, we will register listeners on the bucket. 
                         CompletableFuture<Void> bucketFuture = new CompletableFuture<>();
+
+                        buckets.computeIfAbsent(bucket, x -> bucketServiceSupplier.apply(bucket));
+
+                        AbstractBucketService bucketService = buckets.get(bucket);
                         bucketService.addListener(new Listener() {
```

Comment:
I recommend using `Services` (in `common`) to attach these listeners. They make the code look simpler and handle some odd corner cases (for instance, if the service is already running by the time you attach this listener, the `running` will not execute. Same with `stopped`/`failed`.

EDIT: I see you are starting the service afterwards. That will solve the second problem. But it still provides a nice `Services.startAsync` method which returns a `CompletableFuture` which you can use to invoke the `running` callback.